server/src/main/java/io/deephaven/server/session/SessionService.java

```diff
@@ -70,10 +85,43 @@ public StatusRuntimeException transform(final Throwable err) {
                 }
                 return sre;
             } else if (err instanceof InterruptedException) {
-                return GrpcUtil.securelyWrapError(log, err, Code.UNAVAILABLE);
+                return securelyWrapError(log, err, Code.UNAVAILABLE);
             } else {
-                return GrpcUtil.securelyWrapError(log, err);
+                return securelyWrapError(log, err, Code.INVALID_ARGUMENT);
+            }
+        }
+
+        private synchronized StatusRuntimeException securelyWrapError(
+                final Logger log,
+                final Throwable err,
+                final Code statusCode) {
+            UUID errorId;
+            Throwable curr = err;
+            int currDepth = 0;
+            boolean needToAdd = false;
+            do {
+                errorId = idCache.getIfPresent(curr);
+                needToAdd |= errorId == null;
+            } while (errorId == null && ++currDepth < MAX_STACK_TRACE_CAUSAL_DEPTH && (curr = curr.getCause()) != null);
+
+            if (needToAdd) {
+                if (errorId == null) {
+                    errorId = UuidCreator.getRandomBased();
+                }
+
+                curr = err;
+                do {
+                    if (curr.getStackTrace().length != 0) {
+                        // Stackless exceptions are not very useful, so we only add to the cache if the stack exists.
+                        idCache.put(curr, errorId);
+                    }
```

Comment:
Is this `if` condition meaningful with identity keys? (I can't see how that would be the case.) I think we can safely dump the check and cache unconditionally.

Reply:
Apparently when the jvm starts to throw stackless exceptions, it throws singletons.

```java
    @Test
    public void testThing() {
        int cnt = 0;
        long ttl = 0;
        int[] arr = new int[0];
        Exception[] es = new Exception[10];
        while (true) {
            if (++ttl == 1000000L) break;
            try {
                arr[0] += cnt;
            } catch (ArrayIndexOutOfBoundsException npe) {
                if (npe.getStackTrace().length == 0) {
                    es[cnt++] = npe;
                    if (cnt == es.length) {
                        break;
                    }
                }
            }
        }

        System.out.println("Total Loops: " + ttl);
        for (int i = 1; i < cnt; ++i) {
            System.out.println("es[" + i + "] == es[0] is " + (es[i] == es[0]));
        }
    }
```

Output:
```
Total Loops: 99336
es[1] == es[0] is true
es[2] == es[0] is true
es[3] == es[0] is true
es[4] == es[0] is true
es[5] == es[0] is true
es[6] == es[0] is true
es[7] == es[0] is true
es[8] == es[0] is true
es[9] == es[0] is true
```